### PR TITLE
#10874: Enable test cases for concurrent instances in CCL all gather

### PR DESCRIPTION
**Pull request by Aswinmcw**
Enable test cases for concurrent instances in CCL all gather
Tracking Issue : #10874 

T3K Nightly Test CI - https://github.com/tenstorrent/tt-metal/actions/runs/10483348824
T3K Frequent Test - https://github.com/tenstorrent/tt-metal/actions/runs/10492084492
All Post Commit Test - https://github.com/tenstorrent/tt-metal/actions/runs/10493261845

### Checklist
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by SeanNijjar** [comment]
Posting info from our offline discussion to the PR for reference:
- new line all-gather tests ported from ring all-gather should be moved to a new nightly list for line allgather
- Adopt APIs presented in the PR for new tests https://github.com/tenstorrent/tt-metal/pull/10872

File: tests/sweep_framework/sweeps/line_all_gather.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+import tt_lib as ttl
+import ttnn
+from tests.tt_eager.python_api_testing.sweep_tests.comparison_funcs import comp_equal, comp_pcc
+from models.utility_functions import skip_for_grayskull, get_devices_for_t3000
+
+# Override the default timeout in seconds for hang detection.
+TIMEOUT = 30
+
+
+def is_unsupported_case(input_shape, dim, mem_config, num_devices, num_links, input_dtype, layout):
```
Comment (SeanNijjar) [inline tests/sweep_framework/sweeps/line_all_gather.py:23]: Please import this from the base test_all_gather file if this is a duplicate of that one

File: tests/sweep_framework/sweeps/line_all_gather.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    devices = get_devices_for_t3000(all_devices, num_devices)
+    # for device in devices:
+    #    device.disable_and_clear_program_cache()
+
+    logger.info(f"Input shape: {input_shape}")
+    logger.info(f"dim: {dim}")
+
+    input_tensor = torch.rand(input_shape).bfloat16()
+
+    input_tensors = torch.chunk(input_tensor, num_devices, dim)
```
Comment (SeanNijjar) [inline tests/sweep_framework/sweeps/line_all_gather.py:154]: This is not using the proper `ShardTensor2dMesh` API as outlined in this PR: https://github.com/tenstorrent/tt-metal/pull/10872. Please update this and all future tests to use this API as it's the official recommendation way to describe these tensors now.

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+import pytest
+from loguru import logger
+import tt_lib as ttl
+import ttnn
+from tests.tt_eager.python_api_testing.sweep_tests.comparison_funcs import comp_equal, comp_pcc
+from models.utility_functions import skip_for_grayskull, get_devices_for_t3000
+import itertools
+
+
+def is_unsupported_case(input_shape, dim, mem_config, num_devices, num_links, input_dtype, layout):
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:15]: import from base test_all_gather

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    devices = get_devices_for_t3000(all_devices, num_devices)
+    # for device in devices:
+    #    device.disable_and_clear_program_cache()
+
+    logger.info(f"Input shape: {input_shape}")
+    logger.info(f"dim: {dim}")
+
+    input_tensor = torch.rand(input_shape).bfloat16()
+
+    input_tensors = torch.chunk(input_tensor, num_devices, dim)
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:105]: Same API usage comment

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    "mem_config",
+    [
+        ttl.tensor.MemoryConfig(
+            buffer_type=ttl.tensor.BufferType.DRAM
+        ),  # https://github.com/tenstorrent/tt-metal/issues/9686
+        # ttl.tensor.MemoryConfig(buffer_type=ttl.tensor.BufferType.L1),
+    ],
+)
+@pytest.mark.parametrize("enable_async", [True, False])
+def test_line_all_gather_on_t3000_post_commit(
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:164]: What's the difference between this and the same function in test_all_gather?

Comment (Aswinmcw) [reply tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:107]: This def is using `t3k_device_mesh` fixture instead of old `all_devices`

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    "mem_config",
+    [
+        ttl.tensor.MemoryConfig(
+            buffer_type=ttl.tensor.BufferType.DRAM
+        ),  # https://github.com/tenstorrent/tt-metal/issues/9686
+        # ttl.tensor.MemoryConfig(buffer_type=ttl.tensor.BufferType.L1),
+    ],
+)
+@pytest.mark.parametrize("enable_async", [True, False])
+def test_line_all_gather_on_t3000_post_commit_instances(
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:244]: rename to nightly

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py (hunk 5 above)
Comment (SeanNijjar) [reply tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:107]: Sounds good. Please rename this to "_nightly_"

File: tests/sweep_framework/sweeps/line_all_gather.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    enable_async,
+    num_iters=1,
+) -> list:
+    for device in t3k_device_mesh.get_devices():
+        device.enable_async(enable_async)
+
+    logger.info(f"Input shape: {input_shape}")
+    logger.info(f"dim: {dim}")
+
+    # for device in devices:
```
Comment (SeanNijjar) [inline tests/sweep_framework/sweeps/line_all_gather.py:83]: delete

File: tests/sweep_framework/sweeps/line_all_gather.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    for i, t in enumerate(ttnn.get_device_tensors(tt_out_tensor)):
+        tt_output_tensor = t.cpu().to(ttl.tensor.Layout.ROW_MAJOR).to_torch()
+        if input_dtype == ttl.tensor.DataType.BFLOAT16:
+            eq, output = comp_equal(tt_output_tensor, input_tensor)
+        else:
+            eq, output = comp_pcc(tt_output_tensor, input_tensor)
+        if not eq:
+            logger.error(f"output mismatch for tensor {i}")
+        # assert eq, f"{i} FAILED: {output}"
```
Comment (SeanNijjar) [inline tests/sweep_framework/sweeps/line_all_gather.py:109]: delete

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+    logger.info(f"Input shape: {input_shape}")
+    logger.info(f"dim: {dim}")
+
+    (is_known_failure, message) = is_unsupported_case(
+        input_shape, dim, mem_config, num_devices, num_links, input_dtype, layout
+    )
+    if is_known_failure:
+        pytest.skip(f"Skipping unsupported case {message}.")
+
+    # for device in devices:
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:45]: delete

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+            logger.error(f"output mismatch for tensor {i}")
+        assert eq, f"{i} FAILED: {output}"
+
+
+# Enumerate the post-commit cases explicitly
+@skip_for_grayskull("Requires eth connected devices to run")
+@pytest.mark.parametrize(
+    "num_devices, num_links, input_shape, dim, layout",
+    [
+        # (4, 1, [4, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR), # https://github.com/tenstorrent/tt-metal/issues/9686
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:77]: Please remove the issue comments as they do not apply here.

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+        # ttl.tensor.DataType.BFLOAT8_B, # https://github.com/tenstorrent/tt-metal/issues/9686
+    ],
+)
+@pytest.mark.parametrize(
+    "mem_config",
+    [
+        ttl.tensor.MemoryConfig(
+            buffer_type=ttl.tensor.BufferType.DRAM
+        ),  # https://github.com/tenstorrent/tt-metal/issues/9686
+        # ttl.tensor.MemoryConfig(buffer_type=ttl.tensor.BufferType.L1),
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:103]: Please re-enable this for nightly

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+        # (4, 1, [8, 5, 32, 384], 3, ttl.tensor.Layout.TILE), # https://github.com/tenstorrent/tt-metal/issues/9686
+        (8, 1, [8, 5, 32, 512], 3, ttl.tensor.Layout.TILE),  # https://github.com/tenstorrent/tt-metal/issues/9686
+        (4, 1, [1, 1, 32, 16384], 3, ttl.tensor.Layout.TILE),
+    ],
+)
+@pytest.mark.parametrize(
+    "input_dtype",
+    [
+        ttl.tensor.DataType.BFLOAT16,
+        # ttl.tensor.DataType.BFLOAT8_B, # https://github.com/tenstorrent/tt-metal/issues/9686
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:94]: please re-enable

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+    logger.info(f"Input shape: {input_shape}")
+    logger.info(f"dim: {dim}")
+
+    (is_known_failure, message) = is_unsupported_case(
+        input_shape, dim, mem_config, num_devices, num_links, input_dtype, layout
+    )
+    if is_known_failure:
+        pytest.skip(f"Skipping unsupported case {message}.")
+
+    # devices = get_devices_for_t3000(all_devices, num_devices)
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:167]: delete

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+            if not eq:
+                logger.error(f"output mismatch for tensor {i}")
+            assert eq, f"{i} FAILED: {output}"
+
+
+@skip_for_grayskull("Requires eth connected devices to run")
+@pytest.mark.parametrize(
+    "num_devices, num_instances, num_links, input_shape, dim, layout",
+    [
+        # (4, 1, [4, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR), # https://github.com/tenstorrent/tt-metal/issues/9686
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:213]: please remove the issue part of the comment.

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+        ttl.tensor.DataType.BFLOAT16,
+        # ttl.tensor.DataType.BFLOAT8_B, # https://github.com/tenstorrent/tt-metal/issues/9686
+    ],
+)
+@pytest.mark.parametrize(
+    "mem_config",
+    [
+        ttl.tensor.MemoryConfig(
+            buffer_type=ttl.tensor.BufferType.DRAM
+        ),  # https://github.com/tenstorrent/tt-metal/issues/9686
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:239]: remote issue comment

**Review by SeanNijjar — CHANGES_REQUESTED** [review]
Just minor fixes. Also please the post commit, t3000 unit tests, and t3000 frequent pipelines

File: tests/sweep_framework/sweeps/line_all_gather.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        "dim": [0, 1, 3],
+        "layout": [ttl.tensor.Layout.ROW_MAJOR, ttl.tensor.Layout.TILE],
+        "input_dtype": [ttl.tensor.DataType.BFLOAT16],
+        "mem_config": [ttl.tensor.MemoryConfig(buffer_type=ttl.tensor.BufferType.DRAM)],
+        "enable_async": [True, False],
+    },
+}
+
+
+def skip(
```
Comment (jdesousa-TT) [inline tests/sweep_framework/sweeps/line_all_gather.py:47]: If this is to be used with the new sweep framework, please change this skip function to use the `invalidate_vector` syntax described here: https://docs.google.com/presentation/d/1zo1pS40IZAcVFivAUSJIYljjVerGWJccpVJsP5KL00E/edit#slide=id.g2ec533b0c78_0_82

File: tests/sweep_framework/sweeps/line_all_gather.py [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+    )
+    if is_known_failure:
+        return True, f"Skipping unsupported case {message}."
+    return False, None
+
+
+# This is the run instructions for the test, defined by the developer.
+# The run function must take the above-defined parameters as inputs.
+# The runner will call this run function with each test vector, and the returned results from this function will be stored.
+def run(
```
Comment (jdesousa-TT) [inline tests/sweep_framework/sweeps/line_all_gather.py:63]: Required inputs to this run function must match the defined parameters above, make sure all inputs are defined in the parameters list so the test can run with the sweeps infra.

**Review by jdesousa-TT — CHANGES_REQUESTED** [review]

**Comment by SeanNijjar** [comment]
Just checked with @jdesousa-TT, as of about `e12cs07`, test case generation and reporting should work from within the Austin cloud. Can you please take these changes and generate the new cases and get the initial success rate? Then we can start to merge.

**Review by cfjchu — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:1]: I think this should be `test_all_gather_nightly`- the special line entrypoint will eventually be deleted and all all-gather invocations will go through a generic `all-gather` op (line and ring).

Additionally, the nightly tests from `test_all_gather.py` need to be moved into this file as well.

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+from loguru import logger
+import tt_lib as ttl
+import ttnn
+from tests.tt_eager.python_api_testing.sweep_tests.comparison_funcs import comp_equal, comp_pcc
+from models.utility_functions import skip_for_grayskull, get_devices_for_t3000
+from tests.ttnn.unit_tests.operations.test_all_gather import is_unsupported_case
+from ttnn import ShardTensorToMesh
+
+
+def run_line_all_gather(
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:16]: This run function should probably just go in the base test_all_gather file and replace the corresponding function there. The callers of this function can be updated to use the t3k_device_mesh. Then this should be imported to here.

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+    )
+    if is_known_failure:
+        pytest.skip(f"Skipping unsupported case {message}.")
+
+    t3k_device = []
+
+    for device in t3k_device_mesh.get_devices():
+        t3k_device.append(device)
+
+    t3000_device_rows = [
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:167]: This will need to be updated to use the tensor sharding APIs but we can do that in a separate PR

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+    logger.info(f"Input shape: {input_shape}")
+    logger.info(f"dim: {dim}")
+
+    input_tensor = torch.rand(input_shape).bfloat16()
+
+    ttnn_tensor = ttnn.from_torch(input_tensor, mesh_mapper=ShardTensorToMesh(t3k_device_mesh, dim=dim))
+    input_tensor_mesh = ttnn.to_device(ttnn_tensor, t3k_device_mesh)
+
+    result_mesh_tensors = []
+    for i, devices in enumerate(t3000_device_rows):
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:180]: This loop and the one following should be wrapped in  a loop over `num_iters`

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+        tt_out_tensor = ttnn.line_all_gather(input_tensor_mesh, dim, num_links=num_links, memory_config=mem_config)
+        result_mesh_tensors.append(tt_out_tensor)
+
+    ## Wait for completion
+    for i, devices in enumerate(t3000_device_rows):
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:185]: this is the second loop

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 22]
```diff
@@ -0,0 +1,263 @@
...
+
+@skip_for_grayskull("Requires eth connected devices to run")
+@pytest.mark.parametrize(
+    "num_devices, num_instances, num_links, input_shape, dim, layout",
+    [
+        # (4, 1, [4, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR),
+        # (8, 1, [8, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR),
+        # # (8, 1, [8, 1, 256, 32], 0, ttl.tensor.Layout.TILE),
+        # (8, 1, [8, 8, 256, 384], 1, ttl.tensor.Layout.ROW_MAJOR),
+        # (4, 2, [8, 8, 256, 384], 1, ttl.tensor.Layout.TILE),
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:209]: Is there a reason these other 4-chip cases aren't enabled? The 8chip cases being disabled makes sense (and they should just be removed from the list), but all 4-chip cases here should be enabled.

File: tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+            logger.error(f"output mismatch for tensor {i}")
+        assert eq, f"{i} FAILED: {output}"
+
+
+# Enumerate the post-commit cases explicitly
+@skip_for_grayskull("Requires eth connected devices to run")
+@pytest.mark.parametrize(
+    "num_devices, num_links, input_shape, dim, layout",
+    [
+        # (4, 1, [4, 1, 33, 256], 0, ttl.tensor.Layout.ROW_MAJOR),
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_line_all_gather_nightly.py:74]: why is it commented out?

**Review by SeanNijjar — CHANGES_REQUESTED** [review]

File: tests/ttnn/unit_tests/operations/test_all_gather.py [hunk 24]
```diff
@@ -476,6 +530,71 @@ def test_line_all_gather_on_t3000_post_commit(
...
+        num_iters,
+    )
+
+
+@skip_for_grayskull("Requires eth connected devices to run")
+@pytest.mark.parametrize(
+    "num_devices, num_links, input_shape, dim, layout",
+    [
+        (4, 2, [1, 4, 32, 3584], 1, ttnn.TILE_LAYOUT),
+        # (8, 1, [1, 8, 32, 2048], 1, ttnn.TILE_LAYOUT),
```
[the remark this answers is not in the document]
Comment (Aswinmcw) [reply tests/ttnn/unit_tests/operations/test_all_gather.py]: Done

**Review by SeanNijjar — CHANGES_REQUESTED** [review]
Just need these last couple changes. Please also confirm your changes were picked up by the nightly pipeline. I don't see the necessary change listed here.

Please add the file to the nightly pipeline. You can search for reduce_scatter_nightly invocation to see where to invoke the new test file changes.

Also confirm t3000 frequent and post commit are passing. Post the pipeline links in the PR description.

**Comment by Aswinmcw** [comment]
@SeanNijjar I have added the changes, and i already added this new nightly test to be triggered in ci and also it passed

**Comment by SeanNijjar** [comment]
> @SeanNijjar I have added the changes, and i already added this new nightly test to be triggered in ci and also it passed https://github.com/tenstorrent/tt-metal/actions/runs/10483348824

Please also run post commit (always) and t3000 frequent (since some of the test code was modified for that as well.

**Comment by Aswinmcw** [comment]
> > @SeanNijjar I have added the changes, and i already added this new nightly test to be triggered in ci and also it passed https://github.com/tenstorrent/tt-metal/actions/runs/10483348824
> 
> Please also run post commit (always) and t3000 frequent (since some of the test code was modified for that as well.

Both all post commit test and t3k frequent test are passing, i will ensure remaining tests before merging

**Review by SeanNijjar — APPROVED** [review]

File: tests/sweep_framework/sweeps/line_all_gather.py (hunk 16 above)
Comment (jdesousa-TT) [reply tests/sweep_framework/sweeps/line_all_gather.py:47]: Please update this function to be called `invalidate_vector`

File: tests/sweep_framework/sweeps/line_all_gather.py (hunk 17 above)
Comment (jdesousa-TT) [reply tests/sweep_framework/sweeps/line_all_gather.py:63]: For the device setup, please see this new PR and accompanying README change. This will allow you to define the device mesh you need for your test: https://github.com/tenstorrent/tt-metal/pull/11486

You should copy the device fixture code you require to a generator function called `device_mesh_fixture` in this file. The sweep infra will pick this up, use it, and pass it to the test in the `device` parameter which you should include instead of t3k_device_mesh here.

Comment (Aswinmcw) [reply tests/sweep_framework/sweeps/line_all_gather.py:63]: Thanks for the PR, I will look into it and try adding device generator

**Review by jdesousa-TT — APPROVED** [review]
Looks good to me now. Thanks.

**Review by xanderchin — APPROVED** [review]
